File: srcs/server/orchestration/queue/queue.go
package queue

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"time"

	_ "github.com/mattn/go-sqlite3"
)

type Job struct {
	ID        string          `json:"id"`
	Type      string          `json:"type"`
	Payload   json.RawMessage `json:"payload"`
	Status    string          `json:"status"`
	CreatedAt time.Time       `json:"created_at"`
	UpdatedAt time.Time       `json:"updated_at"`
}

type TaskQueue interface {
	Enqueue(ctx context.Context, job *Job) error
	Dequeue(ctx context.Context, jobTypes []string) (*Job, error)
	Complete(ctx context.Context, jobID string) error
	Fail(ctx context.Context, jobID string, reason string) error
}

type SQLiteTaskQueue struct {
	db *sql.DB
}

func NewSQLiteTaskQueue(db *sql.DB) (*SQLiteTaskQueue, error) {
	_, err := db.Exec(`
		CREATE TABLE IF NOT EXISTS sub_agent_jobs (
			id TEXT PRIMARY KEY,
			type TEXT NOT NULL,
			payload JSON,
			status TEXT NOT NULL,
			created_at DATETIME,
			updated_at DATETIME
		)
	`)
	if err != nil {
		return nil, fmt.Errorf("failed to create jobs table: %w", err)
	}
	return &SQLiteTaskQueue{db: db}, nil
}

func (q *SQLiteTaskQueue) Enqueue(ctx context.Context, job *Job) error {
	now := time.Now()
	_, err := q.db.ExecContext(ctx, `
		INSERT INTO sub_agent_jobs (id, type, payload, status, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?)
	`, job.ID, job.Type, job.Payload, "PENDING", now, now)
	return err
}

func (q *SQLiteTaskQueue) Dequeue(ctx context.Context, jobTypes []string) (*Job, error) {
	tx, err := q.db.BeginTx(ctx, nil)
	if err != nil {
		return nil, err
	}
	defer tx.Rollback()

	var job Job
	var payloadStr string
	var query string
	var args []interface{}

	if len(jobTypes) > 0 {
		query = "SELECT id, type, payload, status, created_at, updated_at FROM sub_agent_jobs WHERE status = 'PENDING' AND type = ? ORDER BY created_at ASC LIMIT 1"
		args = []interface{}{jobTypes[0]}
	} else {
		query = "SELECT id, type, payload, status, created_at, updated_at FROM sub_agent_jobs WHERE status = 'PENDING' ORDER BY created_at ASC LIMIT 1"
	}

	err = tx.QueryRowContext(ctx, query, args...).Scan(&job.ID, &job.Type, &payloadStr, &job.Status, &job.CreatedAt, &job.UpdatedAt)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}

	job.Payload = json.RawMessage(payloadStr)

	_, err = tx.ExecContext(ctx, "UPDATE sub_agent_jobs SET status = 'PROCESSING', updated_at = ? WHERE id = ?", time.Now(), job.ID)
	if err != nil {
		return nil, err
	}

	if err := tx.Commit(); err != nil {
		return nil, err
	}

	job.Status = "PROCESSING"
	return &job, nil
}

func (q *SQLiteTaskQueue) Complete(ctx context.Context, jobID string) error {
	_, err := q.db.ExecContext(ctx, "UPDATE sub_agent_jobs SET status = 'COMPLETED', updated_at = ? WHERE id = ?", time.Now(), jobID)
	return err
}

func (q *SQLiteTaskQueue) Fail(ctx context.Context, jobID string, reason string) error {
	_, err := q.db.ExecContext(ctx, "UPDATE sub_agent_jobs SET status = 'FAILED', updated_at = ? WHERE id = ?", time.Now(), jobID)
	return err
}

type RedisTaskQueue struct {
}

func NewRedisTaskQueue() *RedisTaskQueue {
	return &RedisTaskQueue{}
}

func (q *RedisTaskQueue) Enqueue(ctx context.Context, job *Job) error {
	return nil
}

func (q *RedisTaskQueue) Dequeue(ctx context.Context, jobTypes []string) (*Job, error) {
	return nil, nil
}

func (q *RedisTaskQueue) Complete(ctx context.Context, jobID string) error {
	return nil
}

func (q *RedisTaskQueue) Fail(ctx context.Context, jobID string, reason string) error {
	return nil
}

File: src/server/ohc_jobs/queue.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::{FromRow, PgPool};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Job {
    pub id: Uuid,
    pub tenant_id: String,
    pub job_type: String,
    pub payload: sqlx::types::Json<serde_json::Value>,
    pub status: String,
    pub retry_count: i32,
    pub next_retry_at: DateTime<Utc>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

pub struct JobQueue {
    pool: PgPool,
}

impl JobQueue {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    pub async fn enqueue(
        &self,
        tenant_id: &str,
        job_type: &str,
        payload: serde_json::Value,
    ) -> Result<Job, sqlx::Error> {
        let job: Job = sqlx::query_as!(
            Job,
            r#"
            INSERT INTO ohc_job_queue (tenant_id, job_type, payload)
            VALUES ($1, $2, $3)
            RETURNING id, tenant_id, job_type, payload as "payload: sqlx::types::Json<serde_json::Value>", status, retry_count, next_retry_at, created_at, updated_at
            "#,
            tenant_id,
            job_type,
            sqlx::types::Json(payload) as _,
        )
        .fetch_one(&self.pool)
        .await?;

        Ok(job)
    }

    pub async fn dequeue(&self) -> Result<Option<Job>, sqlx::Error> {
        let job: Option<Job> = sqlx::query_as!(
            Job,
            r#"
            UPDATE ohc_job_queue
            SET status = 'PROCESSING', updated_at = CURRENT_TIMESTAMP
            WHERE id = (
                SELECT id
                FROM ohc_job_queue
                WHERE status = 'PENDING' AND next_retry_at <= CURRENT_TIMESTAMP
                ORDER BY created_at ASC
                FOR UPDATE SKIP LOCKED
                LIMIT 1
            )
            RETURNING id, tenant_id, job_type, payload as "payload: sqlx::types::Json<serde_json::Value>", status, retry_count, next_retry_at, created_at, updated_at
            "#
        )
        .fetch_optional(&self.pool)
        .await?;

        Ok(job)
    }

    pub async fn complete(&self, job_id: Uuid) -> Result<(), sqlx::Error> {
        sqlx::query!(
            r#"
            UPDATE ohc_job_queue
            SET status = 'COMPLETED', updated_at = CURRENT_TIMESTAMP
            WHERE id = $1
            "#,
            job_id
        )
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    pub async fn fail(&self, job_id: Uuid, retry_delay: std::time::Duration) -> Result<(), sqlx::Error> {
        // Calculate next retry time
        // Exponential backoff logic should ideally be applied here based on retry_count
        let next_retry_at = Utc::now() + chrono::Duration::from_std(retry_delay).unwrap_or(chrono::Duration::seconds(60));

        sqlx::query!(
            r#"
            UPDATE ohc_job_queue
            SET status = CASE WHEN retry_count >= 3 THEN 'FAILED' ELSE 'PENDING' END,
                retry_count = retry_count + 1,
                next_retry_at = $2,
                updated_at = CURRENT_TIMESTAMP
            WHERE id = $1
            "#,
            job_id,
            next_retry_at
        )
        .execute(&self.pool)
        .await?;

        Ok(())
    }
}
